src/instruments/src/Common/simVars.tsx

```diff
@@ -0,0 +1,455 @@
+     * retrieve it from the simulator otherwise.
+     * @param name The SimVar to update.
+     * @param unit The unit of the SimVar to update.
+     * @param force Whether to always bypass the cache and always retrieve it
+     * from the simulator.
+     */
+    const retrieve: RetrieveSimVar = (name, unit, force, global) => {
+        const key = getKey(name, unit, global || false);
+        if (cache[key] && !force) {
+            return cache[key].value;
+        }
+
+        const value = global ? SimVar.GetGlobalVarValue(name, unit) : SimVar.GetSimVarValue(name, unit);
+        setCache((oldCache) => ({
+            ...oldCache,
+            [key]: {
+                value,
+                lastUpdatedAgo: 0,
+            },
+        }));
+        return value;
+    }
+
+    /**
+     * This function will be called by the SimVar hooks through the context and
+     * updates the appropriate SimVar for the specific unit with the supplied
+     * value.
+     * @param name The SimVar to update.
+     * @param unit The unit of the SimVar to update.
+     * @param value {*|(function(*): *)} Either the new value for the
+     * SimVar, or an update function that takes the old value and returns an
+     * updated value.
+     * @param proxy If the SimVar used to set the SimVar is different from the
+     * SimVar used to retrieve it, set this parameter to the SimVar for the set
+     * operation.
+     */
+    const update: UpdateSimVar = (name, unit, value, proxy) => {
+        const key = getKey(name, unit, false);
+        setCache((oldCache) => {
+            const newValue = typeof value === 'function' ? value(oldCache[key].value) : value;
+            SimVar.SetSimVarValue((proxy || name), unit, newValue);
+            return {
+                ...oldCache,
+                [key]: {
+                    value: newValue,
+                    lastUpdatedAgo: 0,
+                },
+            };
+        });
+    };
+
+    /**
+     * This function will be called by the useSimVar hook through the context
+     * and ensures the SimVar with the supplied name and unit will be updated
+     * every maxStaleness.
+     */
+    const register: RegisterSimVar = (name, unit, maxStaleness, global) => {
+        const key = getKey(name, unit, global);
+        if (!listeners.current[key]) {
+            listeners.current[key] = [];
+        }
+        listeners.current[key].push(maxStaleness || 0);
+    }
+
+    /**
+     * This function will be called by the useSimVar hook through the context
+     * and notifies us that there is one listener less for this specific SimVar
+     * and unit combination.
+     */
+    const unregister: UnregisterSimVar = (name, unit, maxStaleness, global): void => {
+        const key = getKey(name, unit, global);
+        const old = listeners.current[key];
+        if (!Array.isArray(old) || old.length === 0) {
+            throw new Error('Attempted to unregisterHook with no known listener');
```

Comment:
**thought**: This is a matter of style. I tend to choose idempotency over throwing errors. I'm making the assumption that in general the developer working with these hooks will themselves not be unregistering things (it is a concern of the hooks framework you built). Thus, it would be confusing if internal errors are thrown due to a programming error the developer didn't make.

Reply:
The framework is basically doing reference counting here, and there's an assumption that that is always correct as enforced by useEffect. If we degister more than we register SimVars may stop updating, or we may keep updating SimVars we don't care about if we forget to unregister.

So if we lose track of our count here, that's a really bad thing and it can lead to errors that are subtle or hard-to-debug. I'd rather treat this scenario of an obvious miscount as a critical failure in the system and thus raise an error, rather than just ignoring the situation and leave debugging up to the developer. I think I would like to keep it as-is.

But I agree there is some personal preference involved, the code is written in a slightly defensive manner here.